pkg/mcs/port_set.go

```diff
@@ -0,0 +1,263 @@
+	if ps.IsConflictingWith(other) {
+		return nil, errors.New("Conflicting")
+	}
+
+	ports := make(map[string]mcsv1a1.ServicePort, len(ps.ports))
+	for _, p := range ps.ports {
+		ports[p.Name] = p
+	}
+
+	for _, p := range other.ports {
+		if _, found := ports[p.Name]; !found {
+			ports[p.Name] = p
+		}
+	}
+
+	merged := NewPortSet()
+	for _, p := range ports {
+		merged.ports = append(merged.ports, p)
+	}
+	return merged, nil
+}
+
+// Encodes the port as an annotation. Returns the annotation name and value.
+// An error is returned if either the name or value are invalid.
+func encodePortAsAnnotation(port *mcsv1a1.ServicePort) (string, string, error) {
+	var value string
+
+	key := LIGHTHOUSE_PORT_KEY
+	if port.Name != "" {
+		key = fmt.Sprintf(LIGHTHOUSE_PORT_KEY_NAMED_FORMAT, port.Name)
+	}
+
+	if port.AppProtocol != nil && *port.AppProtocol != "" {
+		value = fmt.Sprintf(LIGHTHOUSE_PORT_VALUE_FORMAT_FULL, port.Port, port.Protocol, *port.AppProtocol)
+	} else if port.Protocol != "" {
+		value = fmt.Sprintf(LIGHTHOUSE_PORT_VALUE_FORMAT_PROTOCOL, port.Port, port.Protocol)
+	} else {
+		value = fmt.Sprintf(LIGHTHOUSE_PORT_VALUE_FORMAT_PORT, port.Port)
+	}
+
+	if err := validation.IsQualifiedName(key); len(err) > 0 {
+		return "", "", errors.New(err[0])
+	} else if err := validation.IsValidLabelValue(value); len(err) > 0 {
+		return "", "", errors.New(err[0])
+	}
+
+	return key, value, nil
+}
+
+// Decodes an annotation to determine if it matches a port specification or not.
+// Returns an error when the annotation name matches but the value can't be decoded.
+// Returns both error and port as nil when the annotation name does not match our
+// expected format.
+func decodePortFromAnnotation(name, value string) (*mcsv1a1.ServicePort, error) {
+	if strings.HasPrefix(name, LIGHTHOUSE_PORT_KEY) {
+		p := &mcsv1a1.ServicePort{}
+
+		if len(name) > len(LIGHTHOUSE_PORT_KEY) {
+			_, err := fmt.Sscanf(name, LIGHTHOUSE_PORT_KEY_NAMED_FORMAT, &p.Name)
+			if err != nil {
+				return nil, errors.New("invalid port name:" + name)
+			}
+		}
+
+		// port spec is formatted as port[-[protocol]-appPort] with [] denoting
+		// optional parts.
+		spec := strings.Split(value, "-")
+		switch len(spec) {
+		case 3:
+			p.AppProtocol = &spec[2]
+			fallthrough
+		case 2:
+			p.Protocol = corev1.Protocol(spec[1])
+			fallthrough
+		case 1:
+			i, err := strconv.Atoi(spec[0])
+			if err != nil {
+				return nil, err
+			}
+			p.Port = int32(i)
+		case 0:
+			return nil, errors.New("invalid port spec:" + value)
+		}
+		return p, nil
+	}
+	return nil, nil
+}
+
+// Equals determines if the two sets have equal definitions for their
+// named ports. Comparison can take into account all ports in the two
+// sets or just the shared set.
+// According to the v1.Service definition only one port can be unnamed
+// and Ports having the same name must agree on the Port, Protocol and
+// AppProtocol.
+func equals(lhs, rhs *PortSet, validateAllPorts bool) bool {
```

Comment:
isnt it a bit confusing that equals() also performs validation ? It should  focus on one thing, perhaps you want to separate into two functions?

Reply:
It might be that the name `validateAllPorts` is confusing here. Its meant to indicate whether to test equality comparing all ports or only ports that are common to the two sets. The function is not doing validation/sanity checks but actually only compares the two PortSets

Reply:
removed completely